src/code/z_player_lib.c

```diff
@@ -1187,21 +2087,275 @@ void Player_DrawCouplesMask(PlayState* play, Player* player) {
+            }
+
+            Matrix_Push();
+            Matrix_Translate(D_801F59B0[i].x, D_801F59B0[i].y, D_801F59B0[i].z, MTXMODE_NEW);
+            Matrix_Scale(phi_f20, temp_f22, phi_f20, MTXMODE_APPLY);
+
+            gSPMatrix(&gfx[0], Matrix_NewMtx(play->state.gfxCtx), G_MTX_PUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+            gSPSegment(&gfx[1], 0x08, VIRTUAL_TO_PHYSICAL(SEGMENTED_TO_VIRTUAL(gEffBubble1Tex)));
+            gDPSetPrimColor(&gfx[2], 0, 0, 255, 255, 255, 255);
+            gDPSetEnvColor(&gfx[3], 150, 150, 150, 0);
+            gSPDisplayList(&gfx[4], gEffBubbleDL);
+            gSPPopMatrix(&gfx[5], G_MTX_MODELVIEW);
+
+            Matrix_Pop();
+
+            gfx = &gfx[6];
+        } else {
+            Player* player = (Player*)actor;
+            f32 temp_f0 = sqrtf(SQ(player->actor.velocity.x) + SQ(player->actor.velocity.z));
+            s16 phi_s0 = temp_f0 * 2000.0f;
+            f32 temp_f20;
+
+            D_801C0BA8.y = temp_f0 * 0.4f;
+            D_801C0BB4.y = -0.3f;
+
+            if (phi_s0 > 0x3E80) {
+                phi_s0 = 0x3E80;
+            }
+
+            phi_s0 = player->actor.focus.rot.y + ((i != 0) ? phi_s0 : -phi_s0);
+            temp_f20 = temp_f0 * 0.2f;
+
+            if (temp_f20 > 4.0f) {
+                temp_f20 = 4.0f;
+            }
+
+            D_801C0BA8.x = -Math_SinS(phi_s0) * temp_f20;
+            D_801C0BA8.z = -Math_CosS(phi_s0) * temp_f20;
+
+            EffectSsDtBubble_SpawnColorProfile(play, &D_801F59B0[i], &D_801C0BA8, &D_801C0BB4, 20, 20, 3, 0);
+            D_801F59C8[i] -= 0x190;
+        }
+    }
+
+    POLY_XLU_DISP = gfx;
+
+    CLOSE_DISPS(play->state.gfxCtx);
+}
+
+void Player_DrawBlastMask(PlayState* play, Player* player) {
+    static Gfx D_801C0BC0[] = {
+        gsDPSetEnvColor(0, 0, 0, 255),
+        gsSPEndDisplayList(),
+    };
+    static Gfx D_801C0BD0[] = {
+        gsDPSetRenderMode(AA_EN | Z_CMP | Z_UPD | IM_RD | CLR_ON_CVG | CVG_DST_WRAP | ZMODE_XLU | FORCE_BL |
+                              G_RM_FOG_SHADE_A,
+                          AA_EN | Z_CMP | Z_UPD | IM_RD | CLR_ON_CVG | CVG_DST_WRAP | ZMODE_XLU | FORCE_BL |
+                              GBL_c2(G_BL_CLR_IN, G_BL_A_IN, G_BL_CLR_MEM, G_BL_1MA)),
+        gsSPEndDisplayList(),
+    };
+
+    OPEN_DISPS(play->state.gfxCtx);
+
+    if (player->unk_B60 != 0) {
+        s32 alpha;
+
+        gSegments[0xA] = VIRTUAL_TO_PHYSICAL(player->maskObjectSegment);
+
+        AnimatedMat_DrawOpa(play, Lib_SegmentedToVirtual(&object_mask_bakuretu_Matanimheader_0011F8));
+
+        if (player->unk_B60 < 11) {
+            alpha = (player->unk_B60 / 10.0f) * 255;
+        } else {
+            alpha = 255;
+        }
+
+        gDPSetEnvColor(POLY_OPA_DISP++, 0, 0, 0, (u8)alpha);
+        gSPDisplayList(POLY_OPA_DISP++, object_mask_bakuretu_DL_000440);
+        gSPSegment(POLY_OPA_DISP++, 0x09, D_801C0BD0);
+        gDPSetEnvColor(POLY_OPA_DISP++, 0, 0, 0, (u8)(255 - alpha));
```

Comment:
I suppose these casts are needed even though they're not actually required.

Reply:
Yes, those are needed 
https://decomp.me/scratch/3oqlc